File: emulation/domain/order.go
package domain

import (
	"errors"
	"log/slog"
)

type partStatus byte

const (
	unknown partStatus = iota
	processing
	rejected
	completed
)

type part struct {
	capacity Capacity
	status   partStatus
}

type Order struct {
	id                OrderId
	tokens            Tokens
	parts             map[CapacityType]*part
	consumerRequest   *ConsumerRequest
	investmentRequest *InvestmentRequest
	cycleCounter      uint
	funded            bool
}

type Score uint

type OrderIdGenerator interface {
	New() OrderId
}

func NewInvestmentOrder(id OrderId, ps ProcessSheet, request InvestmentRequest) *Order {
	parts := make(map[CapacityType]*part, len(ps.Require))
	for t, capacity := range ps.Require {
		parts[t] = &part{capacity, unknown}
	}
	order := &Order{id, 0, parts, nil, &request, 0, false}
	logEvent("order.investment.created",
		withOrderId(id),
		withProducerId(request.ProducerId),
		withProduct(request.Product))
	return order
}

func NewConsumerOrder(id OrderId, ps ProcessSheet, request ConsumerRequest) *Order {
	parts := make(map[CapacityType]*part, len(ps.Require))
	for t, capacity := range ps.Require {
		parts[t] = &part{capacity, unknown}
	}
	order := &Order{id, request.Tokens, parts, &request, nil, 0, true}
	logEvent("order.consumer.created",
		withOrderId(id),
		withConsumerId(request.ConsumerId),
		withProduct(request.Product),
		withTokens(request.Tokens))
	return order
}

type OrderInfo struct {
	Id       OrderId
	Tokens   Tokens
	Required map[CapacityType]Capacity
}

func (i OrderInfo) Fulfilled() bool {
	return len(i.Required) == 0
}

var ErrFundingRequired = errors.New("funding required")

func (o *Order) mustBeFunded() {
	if o.RequiresFunding() {
		panic(ErrFundingRequired)
	}
}

func (o *Order) Info() OrderInfo {
	o.mustBeFunded()
	required := make(map[CapacityType]Capacity, len(o.parts))
	for k, v := range o.parts {
		if v.status == processing || v.status == completed {
			continue
		}
		required[k] = v.capacity
	}
	return OrderInfo{o.id, o.tokens, required}
}

func (o *Order) AgentId() OrderingAgentId {
	o.mustBeFunded()
	if o.investmentRequest != nil {
		return FromProducerId(o.investmentRequest.ProducerId)
	}
	return FromConsumerId(o.consumerRequest.ConsumerId)
}

func (o *Order) RequiresFunding() bool {
	return !o.funded
}

func (o *Order) CutOffPrice() CapacityUnitPrice {
	if o.investmentRequest == nil {
		panic("not an investement order")
	}
	return o.investmentRequest.CutOffPrice
}

func (o *Order) Fund(t Tokens) {
	if !o.RequiresFunding() {
		panic("not in funding status")
	}
	if o.investmentRequest == nil {
		panic("not an investement order")
	}
	o.tokens = t
	o.funded = true
	logEvent("order.funded",
		withOrderId(o.id),
		withTokens(t),
		withProducerId(o.investmentRequest.ProducerId))
}

func (o *Order) getPartStatus(ct CapacityType) partStatus {
	part, ok := o.parts[ct]
	if !ok {
		panic(ErrNotFound)
	}
	return part.status
}

func (o *Order) spendTokens(t Tokens) {
	if o.tokens < t {
		panic("too few tokens left")
	}
	o.tokens -= t
	logEvent("order.tokens.spent",
		withOrderId(o.id),
		withTokens(t),
		slog.Int("remainingTokens", int(o.tokens)))
}

func (o *Order) Processing(ct CapacityType, t Tokens) {
	o.mustBeFunded()
	status := o.getPartStatus(ct)
	if status == processing {
		return
	}
	if status == completed {
		panic(ErrWrongState)
	}
	o.spendTokens(t)
	o.parts[ct].status = processing
	logEvent("order.part.processing",
		withOrderId(o.id),
		withCapacityType(ct),
		withTokens(t))
}

func (o *Order) Completed(ct CapacityType, t Tokens) {
	o.mustBeFunded()
	status := o.getPartStatus(ct)
	if status == completed {
		panic(ErrWrongState)
	}
	if status != processing {
		o.spendTokens(t)
	}
	o.parts[ct].status = completed
	logEvent("order.part.completed",
		withOrderId(o.id),
		withCapacityType(ct),
		withTokens(t))
}

func (o *Order) Rejected(ct CapacityType) {
	o.mustBeFunded()
	status := o.getPartStatus(ct)
	if status == processing || status == completed {
		panic(ErrWrongState)
	}
	o.parts[ct].status = rejected
	logEvent("order.part.rejected",
		withOrderId(o.id),
		withCapacityType(ct))
}

type OrderEvent any
type ConsumerRequestCompleted struct {
	Request *ConsumerRequest
}
type InvestmentRequestCompleted struct {
	Request *InvestmentRequest
}
type ConsumerRequestRejected struct {
	Remaining Tokens
	Request   *ConsumerRequest
}
type InvestmentRequestRejected struct {
	Request *InvestmentRequest
}
type OrderStillProcessing struct {
}

func (o *Order) CompleteCycle() (Score, OrderEvent) {
	o.mustBeFunded()
	rejectedCount := 0
	completedCount := 0
	for ct, part := range o.parts {
		switch part.status {
		case rejected:
			rejectedCount++
			logEvent("order.cycle.part.rejected",
				withOrderId(o.id),
				withCapacityType(ct))
		case completed:
			completedCount++
			logEvent("order.cycle.part.completed",
				withOrderId(o.id),
				withCapacityType(ct))
		}
	}

	logEvent("order.cycle.status",
		withOrderId(o.id),
		slog.Int("completedParts", completedCount),
		slog.Int("rejectedParts", rejectedCount),
		slog.Int("totalParts", len(o.parts)),
		slog.Uint64("cycleCounter", uint64(o.cycleCounter)))

	// completed
	if completedCount == len(o.parts) {
		scores := Score(0)
		if o.consumerRequest != nil {
			logEvent("order.cycle.completed.consumer",
				withOrderId(o.id),
				withConsumerId(o.consumerRequest.ConsumerId),
				withProduct(o.consumerRequest.Product))
			return scores, ConsumerRequestCompleted{o.consumerRequest}
		}
		logEvent("order.cycle.completed.investment",
			withOrderId(o.id),
			withProducerId(o.investmentRequest.ProducerId),
			withProduct(o.investmentRequest.Product))
		return scores, InvestmentRequestCompleted{o.investmentRequest}
	}
	if rejectedCount == len(o.parts) {
		scores := Score(3)
		if o.consumerRequest != nil {
			logEvent("order.cycle.rejected.consumer",
				withOrderId(o.id),
				withConsumerId(o.consumerRequest.ConsumerId),
				withTokens(o.tokens))
			return scores, ConsumerRequestRejected{o.tokens, o.consumerRequest}
		}
		logEvent("order.cycle.rejected.investment",
			withOrderId(o.id),
			withProducerId(o.investmentRequest.ProducerId))
		return scores, InvestmentRequestRejected{o.investmentRequest}
	}
	if o.cycleCounter == 2 {
		scores := Score(5)
		if o.consumerRequest != nil {
			logEvent("order.cycle.timeout.consumer",
				withOrderId(o.id),
				withConsumerId(o.consumerRequest.ConsumerId),
				withTokens(o.tokens))
			return scores, ConsumerRequestRejected{o.tokens, o.consumerRequest}
		}
		logEvent("order.cycle.timeout.investment",
			withOrderId(o.id),
			withProducerId(o.investmentRequest.ProducerId))
		return scores, InvestmentRequestRejected{o.investmentRequest}
	}
	o.cycleCounter++
	logEvent("order.cycle.processing",
		withOrderId(o.id),
		slog.Uint64("cycleCounter", uint64(o.cycleCounter)))
	return Score(1), OrderStillProcessing{}
}

File: emulator/src/domain/order.rs
use std::collections::HashMap;

use super::common::*;

#[derive(PartialEq, Clone, Copy)]
enum PartStatus {
    Completed,
    Processing,
    Rejected,
    Unknown,
}

struct Part {
    capacity: Capacity,
    status: PartStatus,
}

#[derive(Clone)]
enum Investment {
    Restoration,
    Upgrade,
}

#[derive(Clone)]
struct CustomerRequest {
    customer_id: CustomerId,
    product: Product,
    tokens: Tokens,
}

#[derive(Clone)]
struct ProducerRequest {
    producer_id: ProducerId,
    product: Product,
    tokens: Tokens,
    cut_off_price: CapacityUnitPrice,
    investement_type: Investment,
}

#[derive(Clone)]
pub enum Request {
    FromCustomer(CustomerRequest),
    FromProducer(ProducerRequest),
}

#[derive(Hash, Clone, PartialEq, Eq)]
pub struct OrderId(String);
impl OrderId {
    fn new(val: &str) -> Self {
        return OrderId(String::from(val));
    }
}

pub struct Order {
    id: OrderId,
    tokens: Tokens,
    parts: HashMap<CapacityType, Part>,
    request: Request,
    cycle: u8,
}

pub struct OrderInfo {
    id: OrderId,
    tokens: Tokens,
    required: HashMap<CapacityType, Capacity>,
}

impl Order {
    pub fn new(id: OrderId, ps: &ProcessSheet, request: Request) -> Self {
        let parts = ps
            .require
            .iter()
            .map(|(t, capacity)| {
                (
                    *t,
                    Part {
                        capacity: *capacity,
                        status: PartStatus::Unknown,
                    },
                )
            })
            .collect();
        let tokens = match &request {
            Request::FromCustomer(r) => r.tokens,
            Request::FromProducer(r) => r.tokens,
        };
        Self {
            id,
            tokens,
            parts,
            request,
            cycle: 0,
        }
    }

    pub fn info(&self) -> OrderInfo {
        let required = self
            .parts
            .iter()
            .filter(|(_, part)| part.status != PartStatus::Processing && part.status != PartStatus::Completed)
            .map(|(k, v)| (*k, v.capacity))
            .collect();
        return OrderInfo {
            id: self.id.clone(),
            tokens: self.tokens,
            required: required,
        };
    }

    pub fn agent_id(&self) -> OrderingAgentId {
        match &self.request {
            Request::FromCustomer(r) => OrderingAgentId::from_customer(r.customer_id.clone()),
            Request::FromProducer(r) => OrderingAgentId::from_producer(r.producer_id.clone()),
        }
    }

    pub fn requires_funding(&self) -> bool {
        self.tokens.value() == 0
    }

    pub fn cut_off_price(&self) -> CapacityUnitPrice {
        if let Request::FromProducer(r) = &self.request {
            return r.cut_off_price;
        }
        panic!("not an investment order")
    }

    pub fn fund(&mut self, tokens: Tokens) {
        if !self.requires_funding() {
            panic!("not in funding status");
        }
        self.tokens = tokens;
    }

    fn get_part_status(&self, bid: &Bid) -> PartStatus {
        if bid.order_id != self.id {
            panic!("wrong order id");
        }
        self.parts.get(&bid.capacity_type).expect("ErrNotFound").status
    }

    pub fn mark_processing(&mut self, bid: &Bid) {
        match self.get_part_status(bid) {
            PartStatus::Processing => return,
            PartStatus::Completed => panic!("ErrWrongState"),
            _ => {}
        }
        self.tokens.add(bid.tokens);
        self.parts.get_mut(&bid.capacity_type).unwrap().status = PartStatus::Processing;
    }

    pub fn mark_completed(&mut self, bid: &Bid) {
        match self.get_part_status(bid) {
            PartStatus::Processing => {}
            PartStatus::Completed => panic!("ErrWrongState"),
            _ => self.tokens.sub(bid.tokens),
        }
        self.parts.get_mut(&bid.capacity_type).unwrap().status = PartStatus::Completed;
    }

    pub fn makr_rejected(&mut self, bid: &Bid) {
        match self.get_part_status(bid) {
            PartStatus::Processing | PartStatus::Completed => panic!("ErrWrongState"),
            _ => {}
        }
        self.parts.get_mut(&bid.capacity_type).unwrap().status = PartStatus::Rejected;
    }

    pub fn end_cycle(&mut self) -> (Score, OrderEvent) {
        let rejected_count = self.parts.values().filter(|p| p.status == PartStatus::Rejected).count();
        let completed_count = self.parts.values().filter(|p| p.status == PartStatus::Completed).count();
        if completed_count == self.parts.len() {
            let score = Score(0);
            return match &self.request {
                Request::FromCustomer(r) => (score, OrderEvent::CustomerRequestCompleted(r.clone())),
                Request::FromProducer(r) => (score, OrderEvent::ProducerRequestCompleted(r.clone())),
            };
        }
        if rejected_count == self.parts.len() || self.cycle == 3 {
            let score = Score(3);
            return match &self.request {
                Request::FromCustomer(r) => (score, OrderEvent::CustomerRequestRejected(self.tokens, r.clone())),
                Request::FromProducer(r) => (score, OrderEvent::ProducerRequestRejected(self.tokens, r.clone())),
            };
        }
        self.cycle += 1;
        (Score(1), OrderEvent::StillProcessing)
    }
}

pub enum OrderEvent {
    CustomerRequestCompleted(CustomerRequest),
    ProducerRequestCompleted(ProducerRequest),
    CustomerRequestRejected(Tokens, CustomerRequest),
    ProducerRequestRejected(Tokens, ProducerRequest),
    StillProcessing,
}

#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn smoke() {
        let product = Product(1);
        let request = CustomerRequest {
            customer_id: CustomerId("a".to_owned()),
            product: product,
            tokens: Tokens::new(100),
        };
        let ps = ProcessSheet {
            product: product,
            require: HashMap::from([(CapacityType(1), Capacity(100))]),
        };
        let order_id = OrderId::new("abc");
        let mut order = Order::new(order_id.clone(), &ps, Request::FromCustomer(request.clone()));
        assert!(!order.requires_funding());
        let bid = Bid {
            capacity_type: CapacityType(1),
            capacity: Capacity(100),
            tokens: Tokens::new(100),
            order_id: order_id.clone(),
        };
        order.mark_completed(&bid);
        let (score, event) = order.end_cycle();
        assert_eq!(Score(0), score);
        assert!(matches!(event, 
                OrderEvent::CustomerRequestCompleted(CustomerRequest { customer_id, .. }) 
                if customer_id == request.customer_id));
    }
}
